### PR TITLE
Exposing/implementing the HAL driver registry via the C API.

### PR DESCRIPTION
**Pull request by benvanik**
Enables drivers to be provided externally by callers and discovered/loaded dynamically. Zero cost when applications explicitly create their own drivers directly (LTO stripped).

Example use cases are compiling LLVM JIT in its own dylib, remote/sandboxed drivers enumerated based on availability, driver multiversioning, and out-of-tree drivers from downstream projects. Hooray for having a story that will prevent us from ever needing to link CUDA into our binaries ;)

This also let me clean up some of the duplication that had crept in with multiple places all doing driver querying/filtering/etc (and sometimes in slightly different ways). It's intentional that we don't have a non-test utility for that, as non-test targets should not be trying to create all drivers and ignoring any of the metadata the driver registry is now able to provide.

### CODE REVIEW

**Comment by benvanik** [comment]
@KoolJBlack scott mentioned you may be looking at HAL+android stuff soon - this is probably going to be very useful for you :) Happy to gvc and walk through this if you have questions (about this or the HAL API in general).

**Review by ScottTodd — CHANGES_REQUESTED** [review]

File: iree/hal/api.c [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+  // want it to be something dynamic. We could pack an epoch into it that is
+  // bumped each time the registry factory list is modified so we could tell
+  // when a factory was added/removed, etc. So:
+  //   driver_id = [3 byte epoch] [1 byte index into factory list] [4 byte id]
+  // Not sure which status code to return if the epoch is a mismatch, maybe
+  // IREE_STATUS_UNAVAILABLE? If you are mutating the registry from multiple
+  // threads while also enumerating, that may just be enough of a footgun to
+  // bail and force the caller to resolve :)
+  status =
+      iree_make_status(IREE_STATUS_UNIMPLEMENTED, "driver creation by id nyi");
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/hal/api.c:281]: It would - but not now.

File: iree/hal/api.h [hunk 2]
```diff
@@ -1186,29 +1222,137 @@ iree_hal_driver_create_default_device(iree_hal_driver_t* driver,
...
+  // structures (such as the driver name) that callers may choose to clone if
+  // needed.
+  //
+  // Implementers must make their factory enumeration results immutable for the
+  // duration they are registered, though the behavior of try_create is allowed
+  // to change call-to-call. If a factory needs to mutate its set of enumerated
+  // devices then it must do so by first unregistering itself and re-registering
+  // only after the changes have been made.
+  //
+  // Called with the driver registry lock held; may be called from any thread.
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/hal/api.h:1259]: It's not an implementation detail - this is a user-provided function that is getting called with a lock held. It's part of the API. Like with any lock in any situation you can't call a function that also tries to take a lock (so don't call yourself) but can call functions that also expect the lock to be held (like enumerate + try_create).

File: iree/hal/drivers/init.h [hunk 3]
```diff
@@ -29,7 +30,7 @@ extern "C" {
 // create their own user_register_all_available_drivers() that calls this as
 // well as registering their drivers.
 IREE_API_EXPORT iree_status_t IREE_API_CALL
-iree_hal_register_all_available_drivers();
+iree_hal_register_all_available_drivers(iree_hal_driver_registry_t* registry);
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/hal/drivers/init.h:33]: The comment is precise and covers this ("registers all drivers that were linked into the current binary"). If someone doesn't understand the chicken-egg ("I have an empty registry and this method adds things to it! crazy?!") then they are going to have a lot more trouble with this than any comment here will help with.

File: iree/hal/api.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (ScottTodd) [reply iree/hal/api.h:1259]: Ah, I missed some of the "user-provided function" details. Makes more sense seeing how it is used in the `driver_module.cc` files.

**Review by ScottTodd — APPROVED** [review]

**Review by KoolJBlack — APPROVED** [review]
I've looked through this. Would like to have more discussion with Ben, but it looks good. 

Thanks.

**Pull request merged**